pkg/ruler/ruler.go

```diff
@@ -187,26 +188,50 @@ func NewRuler(cfg Config, engine *promql.Engine, queryable promStorage.Queryable
 		logger:       logger,
 	}
 
+	if cfg.EnableSharding {
+		ringStore, err := kv.NewClient(cfg.Ring.KVStore, ring.GetCodec())
+		if err != nil {
+			return nil, errors.Wrap(err, "create KV store client")
+		}
+
+		if enableSharding(ruler, ringStore) != nil {
+			return nil, errors.Wrap(err, "setup ruler sharding ring")
+		}
+	}
+
 	ruler.Service = services.NewBasicService(ruler.starting, ruler.run, ruler.stopping)
 	return ruler, nil
 }
 
+func enableSharding(r *Ruler, ringStore kv.Client) error {
+	lifecyclerCfg, err := r.cfg.Ring.ToLifecyclerConfig()
+	if err != nil {
+		return errors.Wrap(err, "failed to initialize ruler's lifecycler config")
+	}
+
+	// Define lifecycler delegates in reverse order (last to be called defined first because they're
+	// chained via "next delegate").
+	delegate := ring.BasicLifecyclerDelegate(r)
+	delegate = ring.NewLeaveOnStoppingDelegate(delegate, r.logger)
+	delegate = ring.NewAutoForgetDelegate(r.cfg.Ring.HeartbeatTimeout, delegate, r.logger)
```

Comment:
I think this is not correct. I guess you need to apply `ringAutoForgetUnhealthyPeriods` too.

```suggestion
	delegate = ring.NewAutoForgetDelegate(r.cfg.Ring.HeartbeatTimeout * ringAutoForgetUnhealthyPeriods, delegate, r.logger)
```

Reply:
👍 Makes sense. Shouldn't have omitted it